Fix the fur rendering on #4D5307DF

RC0ld opened the pull request:
This PR fixes a rendering issue with 3D models with fur rendered with shell texturing (specifically character skins in games like #4D5307DF rendered as black (issue [#6](https://github.com/xenia-canary/game-compatibility/issues/6)).

BEFORE
<img width="1272" height="707" alt="immagine" src="https://github.com/user-attachments/assets/480cb6b6-4a3e-4ab7-a2d9-eca25cc60f23" />
<img width="1261" height="701" alt="immagine" src="https://github.com/user-attachments/assets/f9c9ea42-0585-4aa0-b928-5215286ae480" />

AFTER
<img width="1264" height="708" alt="immagine" src="https://github.com/user-attachments/assets/da5cc8c4-5d16-4a74-99a1-073341300a7e" />
<img width="1270" height="708" alt="immagine" src="https://github.com/user-attachments/assets/f336e474-d245-4712-ab88-76df6095ec7e" />

I have managed to debug this error because i have thoroughly reverse engineered the texture files this game uses, and came to the realization that for this kind of models where a fur was to be rendered, a 3D texture was loaded. This 3D texture also contained, at its first layer, the texture to be applied to the skin of such models. It is very likely then that the Xenos GPU allows shaders to sample a 3D texture using a 2D sampler (treating it as a 2D array). However, D3D12 does not allow creating a Texture2DArray SRV for a Texture3D resource.

Previously, Xenia attempted to create an invalid SRV or bind the 3D resource to a 2D slot, resulting in no resources being passed to the pixel shader, resulting in this black rendering.

I solved this problem by implementing a "Copy-on-Demand" mechanism in D3D12TextureCache:

- In FindOrCreateTextureDescriptor, we now detect if a shader requests a k2D view for a resource that is actually k3D.

- Instead of creating an invalid view, we call GetOrCreate3DAs2DResource.

- This helper creates a temporary, single-slice Texture2D and issues a CopyTextureRegion command.

- A D3D12_BOX is used with back=1 to ensure only Depth Slice 0 is copied. Copying the full volume to a single slice was previously causing upload buffer corruption.

- The 2D alias is cached in the D3D12Texture object and reused for subsequent frames.

- Added Clear3DAs2DResource to LoadTextureDataFromResidentMemoryImpl to ensure the alias is discarded if the guest game updates the parent 3D texture.


Technical Changes:

- Updated SRVDescriptorKey to include dimension in the hash to prevent collisions between 2D and 3D views of the same asset.

- Added descriptor_key.key = 0 initialization to prevent stack garbage from causing descriptor leaks.


From my testing ( GPU NVIDIA RTX 3060 Ti, game #4D5307DF) the performance impact is negligible.

Triang3l commented:
This is an amazing idea — that was hiding in plain sight — to avoid adding lots of conditionals to translated shaders!

Although I think the implementation can be somewhat simplified by, instead of introducing dependencies between host textures, just loading the texture as a usual 2D host texture, but using 3D tiling while reading the data from the guest memory. That'd be conceptually somewhat similar to the concept of "separate signed version" used in Xenia when, for instance, converting 10bpc and 11bpc guest textures to host 16bpc using different host textures (and copy shaders) for UNorm and SNorm encodings, but here only extent calculations and a boolean in the load shader constants will need to be changed.

One thing that may behave differently between a 3D texture and a 2D view of Z = 0 though is repeating or clamping to border along the Z axis, as with linear filtering, at Z = 0.0, there should be 50% linear interpolation between the first texel and the last texel or the border color in those cases (or some other fraction of the border color with the "clamp to halfway" addressing mode from the ancient OpenGL, though Xenia doesn't support that mode properly at all). So maybe it'd be better to safer to just implement `tfetch2D` and `tfetch3D` the same, but well, that'd add pixel cost overhead to all games, so I'm not even sure really…

RC0ld commented:
Thank you for the review and the kind words!

What i found is quite literally a workaround in all aspects. I would love to refine the idea, following your suggestions. Loading the data directly as a standard 2D host texture but forcing '3D tiling' math during the read could actually be a cleaner way to solving the problem.

I would like to pivot to that solution, but since I am new to the codebase, I could use a small pointer to get started in the right direction.

Regarding the "boolean in the load shader constants": could you point me to the specific struct or location where it would be best to inject this flag? I want to make sure I'm hooking into the existing constant buffer correctly without breaking the standard 2D/3D load paths.

RC0ld commented:
@Triang3l Thanks again for the feedback! I have completely removed the CopyTextureRegion approach and implemented the suggestion to handle this inside the loader itself, reading the guest memory using 3D tiling math but writing directly to a transient 2D host resource.

Also while testing i noticed that this game aggressively aliases memory (reusing the same address for different assets). Xenia's descriptor cache was serving stale descriptors for new assets, causing crashes and potentally visual bugs. So i forced Xenia to refresh the view when the underlying memory changes, fixing the corruption seen during character swaps. So now there should not be anymore problems introduced by my solution with #4D5307DF.

Please let me know if i implemented correctly your suggestion, thank you in advance. If so, i will rebase the commits, so to have a clean history and comply with the contributing guide.

has207 commented:
Can you squash these into a single commit or is there a reason why there are two?

RC0ld commented:
As i tried to express in the previous comment, i tried to implement a different solution to the problem that @Triang3l suggested as more efficient. In that matter i thought that having both commits, with practically two different approaches, could help in understanding if my implementation followed what was in reality suggested to me. And as i said, if @Triang3l, or anyone else on his behalf, can assess that, i will later squash the commits (i used the word rebase as to refer to interactive rebase).

doodshad commented:
Hi everyone, I just downloaded Xenia and started playing this game. I see the issue with the fur being fixed, but I'm not tech savvy so if anyone is alright with explaining how to implement such improvements (top to bottom / dumb it down for me) to this game that be awesome.

RC0ld commented:
> Hi everyone, I just downloaded Xenia and started playing this game. I see the issue with the fur being fixed, but I'm not tech savvy so if anyone is alright with explaining how to implement such improvements (top to bottom / dumb it down for me) to this game that be awesome.

Please for general support refer to the Discord server, let's keep this comment section only related to the PR. Thank you!

Pombaroom commented:
Thank you. Tried that fix - it works, but only with D3D12 API (not working with Vulkan API)
And even with D3D12 API it fixes only world models, but "fury" portraits still look broken, seems like a part of it is transparent:
![4D5307DF - 2025-12-12T23-00-54](https://github.com/user-attachments/assets/3805b7c2-c37f-4ba9-b177-3bada770e517)

antoniodesousa commented:
@RC0ld Your second commit broke something because now after 5 minutes playing all graphics become a mess and crashes after a minute. Also your PR totally breaks Forza Horizon. It doesn't even reach gameplay now after your latest changes.

RC0ld commented:
@Pombaroom Thanks for the feedback. Yes this fix only works with D3D12 API. To my knowledge the game has always had problems with the Vulkan API. 

Regarding the issue that you mentioned, it is not related to the one i fixed, and can be fixed by setting "render_target_path_d3d12" to ROV in the config file.

RC0ld commented:
@antoniodesousa Thanks for the feedback. Please refer to games by their hexadecimal title ID number. Could you please document further where and how are you experiencing crashes and grafical corruption? Provide also your config settings. This will make me easier to debug and fix the issue.

antoniodesousa commented:
> @antoniodesousa Thanks for the feedback. Please refer to games by their hexadecimal title ID number. Could you please document further where and how are you experiencing crashes and grafical corruption? Provide also your config settings. This will make me easier to debug and fix the issue.

Just to be clear, I'm not the only one having these issues.
Several other users reported the same issues in the discord server including @Gliniak.

### 4D5307DF - Blue Dragon
Default settings
Graphical corruption after ~5 minutes playing
Becomes totally black at some point and eventually crashes

### 4D5309C9 - Forza Horizon
Default settings
Crashes before reaching gameplay during the loading screen

CPU: AMD Ryzen 9 6900HX with Radeon Graphics        
OS Platform: Windows
OS Version: 24H2
Manufacturer: ATI Technologies Inc.
GPU Model: AMD Radeon RX 6600M
Driver: 4.0.0 Core Profile Context 25.12.1.251128
Shader: 4.60

RC0ld commented:
@antoniodesousa Thank you again for the feedback. I have made changes to the code that should now fix the issue with #4D5309C9, please if you can test it further. 

Regarding #4D5307DF could you test this one again? Also, in what part of the game does it crash? Is it consistent? In the Xenia config file have you set "clear_memory_page_state" to true?

At this stage i will leave all these commits, once we arrive at a stable build i will merge them.

antoniodesousa commented:
> @antoniodesousa Thank you again for the feedback. I have made changes to the code that should now fix the issue with #4D5309C9, please if you can test it further.

Sadly, it still crashes. It just random now. It can crash within a minute or 10 minutes.
At least it reaches gameplay now.

<img width="512" height="394" alt="LaunchBox 12_16_2025 12_04_13 AM" src="https://github.com/user-attachments/assets/4d009212-5ca5-4409-b166-ec3e566b9de9" />


Blue Dragon seems to be fine now. I only played around 20 minutes.

RC0ld commented:
@antoniodesousa Thank you. I have already identified the logic in the code that causes the crash. I have refined it and it should now be fixed. Any testing by you or others would be greatly appreciated.

antoniodesousa commented:
> @antoniodesousa Thank you. I have already identified the logic in the code that causes the crash. I have refined it and it should now be fixed. Any testing by you or others would be greatly appreciated.

I played around one hour each game. No issues at all. Good work!

antoniodesousa commented:
This was posted in the discord server but Binary Domain crashes immediately when you start a new game or continue where you left off.

### 5345085C - Binary Domain
Default settings
Crashes after start a new game or continue your previous game

### Log file:
[xenia.zip](https://github.com/user-attachments/files/24244828/xenia.zip)

RC0ld commented:
Alright. I performed a significant refactor of the "load texture" logic, because it ultimately became too complicated and convoluted.

Additionally, I identified that this specific texture implementation (copy or load) behaves differently from standard textures regarding the cache. When one of these textures is swapped (aliased at the same memory address), the cache does not detect the change, causing the old texture to persist.

The bug looks like this:

1st model gets loaded
<img width="1282" height="772" alt="immagine" src="https://github.com/user-attachments/assets/b088bc33-c7d5-4d6d-bd41-9b7c8bb2564e" />

2nd model gets loaded
<img width="1282" height="772" alt="immagine" src="https://github.com/user-attachments/assets/7895455b-0fad-4ad9-ae97-34375e200797" />

3rd model gets loaded (Texture mismatch visible)
<img width="1282" height="772" alt="immagine" src="https://github.com/user-attachments/assets/8c5cae6d-e580-409e-b425-66751f157e38" />

3rd model after pressing F5 (Cache cleared, correct texture appears)
<img width="1282" height="772" alt="immagine" src="https://github.com/user-attachments/assets/69b63538-9857-4181-969c-bcac40a75c1e" />

I attempted several methods to solve this, including manual SRV invalidation among other things, but the code became overly complex and started causing regressions in other titles.

Since my free time is coming to an end, I think it is best to leave it at this stage. Perhaps I (or someone with more experience) can revisit the caching issue in the future.

Current PR Structure:

- Commit 1: The original implementation using the "copy texture" logic.
- Commit 2: My attempt to follow @Triang3l's directions using the load logic.

As mentioned many times now, if this second approach will be considered good and well implemented, i will discard the first commit and only keep this one. Otherwise, the first commit remains a viable fallback.

antoniodesousa commented:
@RC0ld I tested both commits and commit 1 works pretty good except for some graphical glitches on the fur textures (yes, I have `clear_memory_page_state = true` and F5 does nothing) and commit 2 has a serious regression in all games including Blue Dragon. They all crash in the first 5 minutes of gameplay. Like a memory leak or something that get's corrupted overtime. On the good side, Binary Domain reaches gameplay now except you only have around 5 minutes to play it before it crashes (like any other game I tried).

### Commit 1 graphical glitch
<img width="1280" height="720" alt="4D5307DF - 2025-12-20T15-52-42" src="https://github.com/user-attachments/assets/018c0b3f-45af-46fb-b040-a3f8aaf2ee68" />

### Commit 2 crash log
[xenia.zip](https://github.com/user-attachments/files/24273480/xenia.zip)

RC0ld commented:
@antoniodesousa Thank you again for your valuable feedback. 

Unfortunately, I am unable to reproduce either the graphics issue from the first commit or the crash from the second, but of course i believe you.

The bug on the first commit is unexpected to me, since it is basically my very first commit, minus some line of codes that triggered a crash rather than a graphical bug during the behaviour that i described in my last comment. 

The crashes on the second commit do not surprise me.

Im working on writing functions for solving the cache issue, i will see what i can get done,

mahoneyd commented:
where can i download this build? The latest build doesn't have the fix. Thank you

has207 commented:
I patched these into xenia-edge and have done a bit of testing, I also added a cvar to be able to easily swap between the two approaches as well as disable this feature in order to test more easily. From what I can tell so far, I haven't really had any crashes with either but FM4 does not get into gameplay at all with this on (both versions). A bunch of other titles I tested seem to work more or less as before. I haven't looked at the caching issue yet, that does need to be addressed I think, though if the feature is opt-in and F5 resolves the immediate issue it's still a better state for Blue Dragon than not having it, the rendering improvements are significant.

has207 commented:
@RC0ld 
I fixed the crashes and rendering issues mentioned above due to how these 2d views were handled in the texture cache (by just not adding them to the cache essentially), also added Vulkan version though Vulkan has some pre-existing texture cache issue but it doesn't seem related to this change. Anyway, it's been in xenia-edge for a few days and seems stable, do you want to pull that into this PR or if you're not planning on working on it anymore I can create a new one integrating your changes here as base. 

The commits you'd need (in order) are ef7d70ff1,  f89f4076, and f776e5e0f then squash with the 2 that are in this PR.

RC0ld commented:
@has207 
That is exactly what i tried to implement but lacked the skills and knowledge necessary. Thank you very much. I will proceed to pull the chages and then squash the commits.

has207 commented:
@RC0ld you will also need 587f29e1, that fixes the logspam with unsupported texture formats logspam

RC0ld committed:
[D3D12] Fix 4D5307DF fur textures

- Fixes issue xenia-canary/game-compatibility#6

AnthonyNiino commented:
> @RC0ldSeu segundo commit quebrou alguma coisa porque agora, depois de 5 minutos de jogo, todos os gráficos ficam uma bagunça e travam depois de um minuto. Além disso, seu PR quebra totalmente o Forza Horizon. Ele nem chega à jogabilidade agora, depois das últimas mudanças.

The same thing happened to me; I had to clear the shader cache and the cache folders, and then it stopped crashing. There were a few graphical bugs, but it's still in the experimental stage. I still saw some bugs, at least in bears. But it still didn't hinder my gameplay; after the village, I didn't see any other glitches, not even in the robot section. I played for 4 hours.

AnthonyNiino commented:
> Este PR corrige um problema de renderização com modelos 3D com pele renderizada com texturização de concha (especificamente skins de personagens em jogos como #4D5307DF renderizadas como pretas (problema [#6](https://github.com/xenia-canary/game-compatibility/issues/6)).
> 
> ANTES <img alt="imaginar" width="1272" height="707" src="https://private-user-images.githubusercontent.com/122227329/517337340-480cb6b6-4a3e-4ab7-a2d9-eca25cc60f23.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3Njk3NTUwMDAsIm5iZiI6MTc2OTc1NDcwMCwicGF0aCI6Ii8xMjIyMjczMjkvNTE3MzM3MzQwLTQ4MGNiNmI2LTRhM2UtNGFiNy1hMmQ5LWVjYTI1Y2M2MGYyMy5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjYwMTMwJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI2MDEzMFQwNjMxNDBaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT1iZDUzNWIyYjM1MTRjNDVkZWVlYTg1YWI0YzFiNDVlZTU2NGJkYTMwZjFlM2E0NWJmMDczZmQ2YTg1ZjZiYjUzJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.bOIt_BmbApYAKOD3XJclE56RKisd3l2EyQwdINpjJRk"> <img alt="imaginar" width="1261" height="701" src="https://private-user-images.githubusercontent.com/122227329/517339644-f9c9ea42-0585-4aa0-b928-5215286ae480.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3Njk3NTUwMDAsIm5iZiI6MTc2OTc1NDcwMCwicGF0aCI6Ii8xMjIyMjczMjkvNTE3MzM5NjQ0LWY5YzllYTQyLTA1ODUtNGFhMC1iOTI4LTUyMTUyODZhZTQ4MC5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjYwMTMwJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI2MDEzMFQwNjMxNDBaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT1jMmQwYzA0OTc2MTUxYmY5ZDM2NGM3ZDg5MjY4MjE4MWI2M2I5ODU5MTg0M2RjOTEzMGNjNjU3MWY5YzM0NmNjJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.3g7nSYsoSvnHZW8ziclZXChzXcoWvhPg0aqHw2XVCgY">
> 
> DEPOIS <img alt="imaginar" width="1264" height="708" src="https://private-user-images.githubusercontent.com/122227329/517340380-da5cc8c4-5d16-4a74-99a1-073341300a7e.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3Njk3NTUwMDAsIm5iZiI6MTc2OTc1NDcwMCwicGF0aCI6Ii8xMjIyMjczMjkvNTE3MzQwMzgwLWRhNWNjOGM0LTVkMTYtNGE3NC05OWExLTA3MzM0MTMwMGE3ZS5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjYwMTMwJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI2MDEzMFQwNjMxNDBaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT03NWEwODcwNzhkYTA1MzhhMTE4NTY0OWQ3YmQ2MzY4YjUyZTEzNWNlNTBhZDYwNzdlMWY1NDQxNWNmNDI3NWJiJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.6DSoWK4m-iFL867zSmtmIZ2nJUKS3gctk5a277KKvWI"> <img alt="imaginar" width="1270" height="708" src="https://private-user-images.githubusercontent.com/122227329/517340806-f336e474-d245-4712-ab88-76df6095ec7e.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3Njk3NTUwMDAsIm5iZiI6MTc2OTc1NDcwMCwicGF0aCI6Ii8xMjIyMjczMjkvNTE3MzQwODA2LWYzMzZlNDc0LWQyNDUtNDcxMi1hYjg4LTc2ZGY2MDk1ZWM3ZS5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjYwMTMwJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI2MDEzMFQwNjMxNDBaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT1hNzdhNmExMTQzNjNiZGM4YzUyNDQ0ZjExZWY0ZWNmMWQxNzI1NzM5ZjM1MGIyMWM5ZjQ4MzA2M2FiZWFmZWM3JlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.3VNLFv0aOOXQ5Wzs8kU_5aJ_ggEkV_I99lvCwo7ChoI">
> 
> Consegui depurar esse erro porque fiz uma engenharia reversa completa dos arquivos de textura que o jogo usa e percebi que, para esse tipo de modelo em que um pelo seria renderizado, uma textura 3D seria carregada. Essa textura 3D também continha, em sua primeira camada, a textura a ser aplicada na pele desses modelos. É muito provável então que a GPU Xenos permita que os shaders amostrem uma textura 3D usando um amostrador 2D (tratando-o como uma matriz 2D). Entretanto, o D3D12 não permite a criação de um Texture2DArray SRV para um recurso Texture3D.
> 
> Anteriormente, o Xenia tentava criar um SRV inválido ou vincular o recurso 3D a um slot 2D, resultando em nenhum recurso sendo passado para o pixel shader, resultando nessa renderização preta.
> 
> Resolvi esse problema implementando um mecanismo "Copiar sob demanda" no D3D12TextureCache:
> 
> * No FindOrCreateTextureDescriptor, agora detectamos se um shader solicita uma visualização k2D para um recurso que na verdade é k3D.
> * Em vez de criar uma visualização inválida, chamamos GetOrCreate3DAs2DResource.
> * Este auxiliar cria um Texture2D temporário de fatia única e emite um comando CopyTextureRegion.
> * Um D3D12_BOX é usado com back=1 para garantir que apenas Depth Slice 0 seja copiado. Copiar o volume completo para uma única fatia estava causando corrupção no buffer de upload.
> * O alias 2D é armazenado em cache no objeto D3D12Texture e reutilizado para quadros subsequentes.
> * Adicionado Clear3DAs2DResource ao LoadTextureDataFromResidentMemoryImpl para garantir que o alias seja descartado se o jogo convidado atualizar a textura 3D pai.
> 
> Mudanças técnicas:
> 
> * SRVDescriptorKey atualizado para incluir dimensão no hash para evitar colisões entre visualizações 2D e 3D do mesmo ativo.
> * Adicionada inicialização descriptor_key.key = 0 para evitar que o lixo da pilha cause vazamentos no descritor.
> 
> Pelos meus testes (GPU NVIDIA RTX 3060 Ti, jogo #4D5307DF) o impacto no desempenho é insignificante.
Keep up the good work on the project 🙏

AnthonyNiino commented:
One problem I was having was that the game would freeze after a few minutes, like 5 minutes, and the emulator would give an error. The only way I tried to fix it was by deleting the shader cache and all the other cache folders on Xenia. That way I was able to play without any more problems with crashes. Until then... It's possible that later in the game there will be texture errors again, or not. I played for about 4 hours, reached the desert and went underground after the fight with Marumaru.

has207 commented:
Just a heads up, there is another bug with the 3d-to-2d code, I'm testing a fix and plan to merge into canary at some point soon so no need to keep updating this PR, I will merge your changes once the issue is resolved

has207 commented:
The original changes from this PR are now merged, still missing fixes to the "portrait" rendering that are present on edge, those will be merged after bytecode generation changes since they affect the precompiled resolve shaders
